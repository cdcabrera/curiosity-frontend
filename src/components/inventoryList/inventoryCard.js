import React from 'react';
import PropTypes from 'prop-types';
import { Bullseye, Card, CardActions, CardBody, CardFooter, CardHeader, CardHeaderMain } from '@patternfly/react-core';
import { connect, reduxActions, reduxSelectors } from '../../redux';
import { MinHeight } from '../minHeight/minHeight';
import Pagination from '../pagination/pagination';
import { ToolbarFieldDisplayName } from '../toolbar/toolbarFieldDisplayName';
import { paginationHelpers } from '../pagination/paginationHelpers';
import { translate } from '../i18n/i18n';

/**
 * A hosts system inventory component.
 *
 * @fires onColumnSort
 * @fires onPage
 * @param {object} props
 * @param {boolean} props.isDisabled
 * @param {boolean} props.error
 * @param {boolean} props.fulfilled
 * @param {number} props.itemCount
 * @param {boolean} props.pending
 * @param {Function} props.t
 * @returns {Node}
 */
const InventoryList = ({
  isDisabled,
  error,
  fulfilled,
  itemCount,
  pending,
  t,
  updatedOffset,
  updatedPerPage,
  onPage,
  children
}) => {
  if (isDisabled) {
    return (
      <Card className="curiosity-inventory-card__disabled">
        <CardBody>
          <Bullseye>{t('curiosity-inventory.tab', { context: 'disabled' })}</Bullseye>
        </CardBody>
      </Card>
    );
  }

  const isLastPage = paginationHelpers.isLastPage(updatedOffset, updatedPerPage, itemCount);

  // Set an updated key to force refresh minHeight
  const minHeightContentRefreshKey =
    (fulfilled === true && itemCount < updatedPerPage && `bodyMinHeight-${updatedPerPage}-resize`) ||
    (fulfilled === true && isLastPage && `bodyMinHeight-${updatedPerPage}-resize`) ||
    (error === true && `bodyMinHeight-${updatedPerPage}-resize`) ||
    `bodyMinHeight-${updatedPerPage}`;

  return (
    <Card className="curiosity-inventory-card">
      <MinHeight key="headerMinHeight" updateOnContent>
        <CardHeader className={(error && 'hidden') || ''} aria-hidden={error || false}>
          <CardHeaderMain>
            <ToolbarFieldDisplayName />
          </CardHeaderMain>
          <CardActions className={(!itemCount && 'transparent') || ''} aria-hidden={!itemCount || false}>
            <Pagination
              isCompact
              isDisabled={pending || error}
              itemCount={itemCount}
              offset={updatedOffset}
              onPage={onPage}
              onPerPage={onPage}
              perPage={updatedPerPage}
            />
          </CardActions>
        </CardHeader>
      </MinHeight>
      <MinHeight key={minHeightContentRefreshKey} updateOnContent>
        <CardBody>
          <div className={(error && 'blur') || 'fadein'}>{children}</div>
        </CardBody>
      </MinHeight>
      <MinHeight key="footerMinHeight" updateOnContent>
        <CardFooter
          className={(error && 'hidden') || (!itemCount && 'transparent') || ''}
          aria-hidden={error || !itemCount || false}
        >
          <Pagination
            dropDirection="up"
            isDisabled={pending || error}
            itemCount={itemCount}
            offset={updatedOffset}
            onPage={onPage}
            onPerPage={onPage}
            perPage={updatedPerPage}
          />
        </CardFooter>
      </MinHeight>
    </Card>
  );
};

/**
 * Prop types.
 *
 * @type {{t: Function, listData: Array, session: object, pending: boolean, fulfilled: boolean,
 *     getHostsInventory: Function, perPageDefault: number, isDisabled: boolean, error: boolean,
 *     itemCount: number}}
 */
InventoryList.propTypes = {
  error: PropTypes.bool,
  fulfilled: PropTypes.bool,
  isDisabled: PropTypes.bool,
  itemCount: PropTypes.number,
  pending: PropTypes.bool,
  t: PropTypes.func
};

/**
 * Default props.
 *
 * @type {{t: Function, listData: Array, session: object, pending: boolean, fulfilled: boolean,
 *     getHostsInventory: Function, perPageDefault: number, isDisabled: boolean, error: boolean,
 *     itemCount: number}}
 */
InventoryList.defaultProps = {
  error: false,
  fulfilled: false,
  isDisabled: false,
  itemCount: 0,
  pending: false,
  t: translate
};

/**
 * Create a selector from applied state, props.
 *
 * @type {Function}
 */
const makeMapStateToProps = reduxSelectors.inventoryList.makeInventoryList();

const ConnectedInventoryList = connect(makeMapStateToProps, mapDispatchToProps)(InventoryList);

export { ConnectedInventoryList as default, ConnectedInventoryList, InventoryList };
