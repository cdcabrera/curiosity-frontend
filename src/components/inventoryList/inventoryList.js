import React from 'react';
import PropTypes from 'prop-types';
import _isEqual from 'lodash/isEqual';
import { TableVariant } from '@patternfly/react-table';
import { helpers } from '../../common';
import { connect, reduxActions, reduxSelectors } from '../../redux';
import Table from '../table/table';
import { Loader } from '../loader/loader';
import GuestsList from '../guestsList/guestsList';
import { translate } from '../i18n/i18n';

class InventoryList extends React.Component {
  componentDidMount() {
    this.onUpdateInventoryData();
  }

  componentDidUpdate(prevProps) {
    const { listQuery, productId } = this.props;

    if (productId !== prevProps.productId || !_isEqual(listQuery, prevProps.listQuery)) {
      this.onUpdateInventoryData();
    }
  }

  onUpdateInventoryData = () => {
    const { getHostsInventory, listQuery, isDisabled, productId } = this.props;

    if (!isDisabled && productId) {
      getHostsInventory(productId, listQuery);
    }
  };

  renderTable() {
    const { listData, listQuery, t } = this.props;

    const rows = listData.map(
      ({ cores, displayName, hardwareType, lastSeen, numberOfGuests, sockets, subscriptionManagerId }) => ({
        cells: [displayName, hardwareType, `${sockets}/${cores}`, lastSeen],
        expandedContent: (
          <GuestsList listQuery={listQuery} listQueryId={subscriptionManagerId} loaderRowCount={numberOfGuests} />
        )
      })
    );

    return (
      <Table
        borders
        variant={TableVariant.compact}
        className="curiosity-inventory-list"
        columnHeaders={[
          t('curiosity-inventory.headerName'),
          t('curiosity-inventory.headerInfrastructure'),
          t('curiosity-inventory.headerSocketsCores'),
          t('curiosity-inventory.headerLastSeen')
        ]}
        rows={rows}
      />
    );
  }

  render() {
    const { error, pending } = this.props;

    return (
      <div className={`curiosity-inventory-list-wrapper fadein ${(error && 'blur') || ''}`}>
        {pending && (
          <Loader
            variant="table"
            tableProps={{
              className: 'curiosity-inventory-list',
              colCount: 4,
              rowCount: 5,
              variant: TableVariant.compact
            }}
          />
        )}
        {!pending && this.renderTable()}
      </div>
    );
  }
}

InventoryList.propTypes = {
  error: PropTypes.bool,
  getHostsInventory: PropTypes.func,
  isDisabled: PropTypes.bool,
  listData: PropTypes.array,
  pending: PropTypes.bool,
  productId: PropTypes.string.isRequired,
  listQuery: PropTypes.object.isRequired,
  t: PropTypes.func,
  viewId: PropTypes.string
};

InventoryList.defaultProps = {
  error: false,
  getHostsInventory: helpers.noop,
  listData: [],
  isDisabled: helpers.UI_DISABLED_TABLE,
  pending: false,
  t: translate,
  viewId: 'inventoryList'
};

/**
 * Apply actions to props.
 *
 * @param {Function} dispatch
 * @returns {object}
 */
const mapDispatchToProps = dispatch => ({
  getHostsInventory: (id, query) => dispatch(reduxActions.rhsm.getHostsInventory(id, query))
});

/**
 * Create a selector from applied state, props.
 *
 * @type {Function}
 */
const makeMapStateToProps = reduxSelectors.inventoryList.makeInventoryList();

const ConnectedInventoryList = connect(makeMapStateToProps, mapDispatchToProps)(InventoryList);

export { ConnectedInventoryList as default, ConnectedInventoryList, InventoryList };
