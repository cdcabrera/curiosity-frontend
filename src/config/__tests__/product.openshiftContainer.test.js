import { config } from '../product.openshiftContainer';
import { parseRowCellsListData } from '../../components/inventoryList/inventoryListHelpers';
import {
  RHSM_API_QUERY_SORT_DIRECTION_TYPES as SORT_DIRECTION_TYPES,
  RHSM_API_QUERY_TYPES
} from '../../types/rhsmApiTypes';

describe('Product OpenShift Container config', () => {
  it('should set product configuration', () => {
    const {
      initialGraphFilters,
      initialInventoryFilters,
      inventoryHostsQuery,
      inventorySubscriptionsQuery,
      query
    } = config;

    expect({ initialGraphFilters, initialInventoryFilters, query }).toMatchSnapshot('initial configuration');

    // filter inventory data checks
    const inventoryData = {
      displayName: 'lorem',
      inventoryId: 'lorem inventory id',
      coreHours: 12.53,
      lastSeen: 'lorem date obj',
      loremIpsum: 'hello world'
    };

    const filteredInventoryData = parseRowCellsListData({
      filters: initialInventoryFilters,
      cellData: inventoryData
    });

    expect(filteredInventoryData).toMatchSnapshot('filteredInventoryData results');

    const fallbackInventoryData = {
      ...inventoryData,
      coreHours: null,
      inventoryId: null,
      lastSeen: null
    };

    const fallbackFilteredInventoryData = parseRowCellsListData({
      filters: initialInventoryFilters,
      cellData: fallbackInventoryData
    });

    expect(fallbackFilteredInventoryData).toMatchSnapshot('filteredInventoryData results, fallback display');

    expect({
      hostsInventory: inventoryHostsQuery[RHSM_API_QUERY_TYPES.DIRECTION] === SORT_DIRECTION_TYPES.DESCENDING,
      subscriptionsInventory:
        inventorySubscriptionsQuery[RHSM_API_QUERY_TYPES.DIRECTION] === SORT_DIRECTION_TYPES.DESCENDING
    }).toMatchSnapshot('default sort for inventory should descend');
  });
});
