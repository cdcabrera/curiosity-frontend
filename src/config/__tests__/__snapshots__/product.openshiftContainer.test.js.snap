// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product OpenShift Container config should set product configuration: default sort for inventory should descend 1`] = `
Object {
  "hostsInventory": true,
  "subscriptionsInventory": true,
}
`;

exports[`Product OpenShift Container config should set product configuration: filteredInventoryData results 1`] = `
Object {
  "cells": Array [
    Object {
      "title": <React.Fragment>
        lorem
         
        
      </React.Fragment>,
    },
    Object {
      "title": "",
    },
    Object {
      "title": "",
    },
    Object {
      "title": <DateFormat
        date="lorem date obj"
      />,
    },
  ],
  "columnHeaders": Array [
    Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"displayName\\"})",
      "transforms": Array [],
    },
    Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"sockets_OpenShift Container Platform\\"})",
      "transforms": Array [
        [Function],
      ],
    },
    Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"cores_OpenShift Container Platform\\"})",
      "transforms": Array [
        [Function],
      ],
    },
    Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"lastSeen\\"})",
      "transforms": Array [
        [Function],
      ],
    },
  ],
  "data": Object {
    "coreHours": Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"coreHours\\"})",
      "value": 12.53,
    },
    "displayName": Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"displayName\\"})",
      "value": "lorem",
    },
    "inventoryId": Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"inventoryId\\"})",
      "value": "lorem inventory id",
    },
    "lastSeen": Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"lastSeen\\"})",
      "value": "lorem date obj",
    },
    "loremIpsum": Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"loremIpsum\\"})",
      "value": "hello world",
    },
  },
}
`;

exports[`Product OpenShift Container config should set product configuration: filteredInventoryData results, fallback display 1`] = `
Object {
  "cells": Array [
    Object {
      "title": "lorem",
    },
    Object {
      "title": "",
    },
    Object {
      "title": "",
    },
    [Function],
  ],
  "columnHeaders": Array [
    Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"displayName\\"})",
      "transforms": Array [],
    },
    Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"sockets_OpenShift Container Platform\\"})",
      "transforms": Array [
        [Function],
      ],
    },
    Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"cores_OpenShift Container Platform\\"})",
      "transforms": Array [
        [Function],
      ],
    },
    Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"lastSeen\\"})",
      "transforms": Array [
        [Function],
      ],
    },
  ],
  "data": Object {
    "coreHours": Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"coreHours\\"})",
      "value": null,
    },
    "displayName": Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"displayName\\"})",
      "value": "lorem",
    },
    "inventoryId": Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"inventoryId\\"})",
      "value": null,
    },
    "lastSeen": Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"lastSeen\\"})",
      "value": null,
    },
    "loremIpsum": Object {
      "title": "t(curiosity-inventory.header, {\\"context\\":\\"loremIpsum\\"})",
      "value": "hello world",
    },
  },
}
`;

exports[`Product OpenShift Container config should set product configuration: initial configuration 1`] = `
Object {
  "initialGraphFilters": Array [
    Object {
      "color": "#06c",
      "fill": "#8bc1f7",
      "id": "cores",
      "isOptional": true,
      "stroke": "#06c",
    },
    Object {
      "color": "#06c",
      "fill": "#8bc1f7",
      "id": "sockets",
      "isOptional": true,
      "stroke": "#06c",
    },
    Object {
      "id": "thresholdSockets",
      "isOptional": true,
    },
    Object {
      "id": "thresholdCores",
      "isOptional": true,
    },
  ],
  "initialInventoryFilters": Array [
    Object {
      "cell": [Function],
      "id": "displayName",
      "isSortable": true,
    },
    Object {
      "cellWidth": 15,
      "header": "t(curiosity-inventory.header, {\\"context\\":\\"sockets_OpenShift Container Platform\\"})",
      "id": "sockets",
      "isOptional": true,
      "isSortable": true,
      "isWrappable": true,
    },
    Object {
      "cellWidth": 15,
      "header": "t(curiosity-inventory.header, {\\"context\\":\\"cores_OpenShift Container Platform\\"})",
      "id": "cores",
      "isOptional": true,
      "isSortable": true,
      "isWrappable": true,
    },
    Object {
      "cell": [Function],
      "cellWidth": 25,
      "id": "lastSeen",
      "isSortable": true,
      "isWrappable": true,
    },
  ],
  "query": Object {
    "beginning": "2019-06-20T00:00:00.000Z",
    "ending": "2019-07-20T23:59:59.999Z",
    "uom": "cores",
  },
}
`;
